nixos/modules/services/web-apps/firefly-iii.nix

```diff
@@ -0,0 +1,356 @@
+      ADLDAP_CONNECTION_SCHEME
+      ADLDAP_CONTROLLERS
+      ADLDAP_PORT
+      ADLDAP_BASEDN
+      ADLDAP_ADMIN_USERNAME
+      ADLDAP_ADMIN_PASSWORD
+      ADLDAP_ACCOUNT_PREFIX
+      ADLDAP_ACCOUNT_SUFFIX
+      WINDOWS_SSO_ENABLED
+      WINDOWS_SSO_DISCOVER
+      WINDOWS_SSO_KEY
+      ADLDAP_SYNC_FIELD
+      TRACKER_SITE_ID
+      TRACKER_URL
+    )
+
+    for e in "''${envs[@]}"; do
+      file_env "$e"
+    done
+  '';
+
+  firefly-iii-maintenance = pkgs.writeScript "firefly-iii-maintenance.sh" ''
+    #!${pkgs.runtimeShell}
+
+    source ${cfg.package}/.env
+
+    ${fileenv-func}
+
+    ${firefly-php} ${artisan} firefly-iii:create-database
+    ${firefly-php} ${artisan} migrate --seed --no-interaction --force
+    ${firefly-php} ${artisan} firefly-iii:decrypt-all
+    ${firefly-php} ${artisan} cache:clear
+    ${firefly-php} ${artisan} firefly-iii:upgrade-database
+    ${firefly-php} ${artisan} firefly-iii:correct-database
+    ${firefly-php} ${artisan} firefly-iii:report-integrity
+    ${firefly-php} ${artisan} firefly-iii:laravel-passport-keys
+    ${firefly-php} ${artisan} cache:clear
+
+    mv /tmp/firefly-iii-env.conf /run/phpfpm/firefly-iii-env.conf
+
+    exit 0
+  '';
+
+  commonServiceConfig = {
+    Type = "oneshot";
+    User = user;
+    Group = group;
+    StateDirectory = "firefly-iii";
+    WorkingDirectory = "${cfg.package}";
+    PrivateTmp = true;
+    PrivateDevices = true;
+    CapabilityBoundingSet = "";
+    AmbientCapabilities = "";
+    ProtectSystem = "strict";
+    ProtectKernelTunables = true;
+    ProtectKernelModules = true;
+    ProtectControlGroups = true;
+    ProtectClock = true;
+    ProtectHostname = true;
+    ProtectHome = "tmpfs";
+    ProtectKernelLogs = true;
+    ProtectProc = "invisible";
+    ProcSubset = "pid";
+    PrivateNetwork = false;
+    RestrictAddressFamilies = "AF_INET AF_INET6 AF_UNIX";
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [
+      "@system-service @resources"
+      "~@obsolete @privileged"
+    ];
+    RestrictSUIDSGID = true;
+    RemoveIPC = true;
+    NoNewPrivileges = true;
+    RestrictRealtime = true;
+    RestrictNamespaces = true;
+    LockPersonality = true;
+    PrivateUsers = true;
+  };
+
+in {
+
+  options.services.firefly-iii = {
+
+    enable = mkEnableOption "Firefly III: A free and open source personal finance manager";
+
+    dataDir = mkOption {
+      description = mdDoc "Firefly III data directory";
+      default = "/var/lib/firefly-iii";
+      type = path;
+    };
+
+    user = mkOption {
+      type = str;
+      default = defaultUser;
+      description = mdDoc "User account under which firefly-iii runs.";
+    };
+
+    group = mkOption {
+      type = str;
+      default = defaultGroup;
+      description = mdDoc ''
+        Group under which firefly-iii runs. It is best to set this to the group
+        of whatever webserver is being used as the frontend.
+      '';
+    };
+
+    package = mkOption {
```

Comment:
Please use mkPackageOption

Reply:
While I've changed it to use `mkPackageOption`, I've not yet found there to be a good method of passing something like `cfg.dataDir` to the default package there. I'd appreciate some guidance on how to do it given I'd like the default package to be `pkgs.firefly-iii.override { dataDir = cfg.dataDir; }`.

Reply:
You cann add apply to the option to do that. eg https://github.com/NixOS/nixpkgs/blob/f59f29659e1db1130f87936ef9f2a11a71b3a99c/nixos/modules/programs/steam.nix#L45

Reply:
This doesn't use `mkPackageOption`. I'll change it to use `mkOption` as per your example.